mission-control/blog/state-based-alerting/index.md

```diff
@@ -0,0 +1,500 @@
+
+### Comparing Alerts: State-Based vs. Prometheus
+
+Now, compare how different monitoring approaches handle this failure.
+
+#### Prometheus Alert (Metric-Based)
+
+With Prometheus, a common alert rule for deployment issues checks for generation mismatches:
+
+```yaml title=KubernetesDeploymentGenerationMismatch
+  - alert: KubernetesDeploymentGenerationMismatch
+    expr: kube_deployment_status_observed_generation != kube_deployment_metadata_generation
+    for: 10m
+    labels:
+      severity: critical
+    annotations:
+      summary: Kubernetes Deployment generation mismatch (instance {{ $labels.instance }})
+      description: "Deployment {{ $labels.namespace }}/{{ $labels.deployment }} has failed but has not been rolled back.\n  VALUE = {{ $value }}\n  LABELS = {{ $labels }}"
+```
+
+This alert fires when there is a mismatch between the observed and expected generation numbers of a Kubernetes Deployment. The generation number increments whenever the Deployment spec changes. A mismatch indicates that the latest configuration change has not been successfully rolled out by the controller. While useful, it doesn't explain *why* the rollout failed. See [KubernetesDeploymentGenerationMismatch](https://samber.github.io/awesome-prometheus-alerts/rules#rule-kubernetes-1-24) for more details on this type of alert.
+
+
+#### Mission Control State-Based Alert
+
+Use `kubectl` and `jq` to inspect the `conditions` within the status of the `failing-deployment`:
+
+```bash
+kubectl get deployment failing-deployment -o json | jq '.status.conditions'
+```
+
+The output might resemble the following:
+
+```json
+[
+  {
+    "lastTransitionTime": "2023-06-15T15:10:23Z",
+    "lastUpdateTime": "2023-06-15T15:10:23Z",
+    "message": "Deployment does not have minimum availability.",
+    "reason": "MinimumReplicasUnavailable",
+    "status": "False",
+    "type": "Available"
+  },
+  {
+    "lastTransitionTime": "2023-06-15T15:15:45Z",
+    "lastUpdateTime": "2023-06-15T15:15:45Z",
+    "message": "ReplicaSet \"failing-deployment-75d55d96c\" has timed out progressing. 0/3 replicas available. Pods are being killed due to memory limit exceeded.",
+    "reason": "ProgressDeadlineExceeded",
+    "status": "False",
+    "type": "Progressing"
+  }
+]
+```
+This output shows two conditions:
+1. `Available` is `False` because the deployment does not have the minimum required replicas ready (`MinimumReplicasUnavailable`).
+2. `Progressing` is `False` because the rollout timed out (`ProgressDeadlineExceeded`). The message provides specific details about the failure, potentially including reasons like OOM killing if the system surfaces that information here.
+
+Mission Control captures this state and provides an alert with the error message from the `Progressing` condition (e.g., "ReplicaSet ... has timed out progressing..."). This points more directly to the root cause or the symptom reported by Kubernetes.
+
```

Comment:
Should we mention somewhere that mission-control supports both ? Maybe in the conclusion

Reply:
It looks like we're comparing with prometheus rules, but we also write about how it is to best use both in the end